### PR TITLE
[WASI-NN] ggml: add llava support

### PR DESCRIPTION
**Pull request by dm4**
- Add `mmproj` and `image` options.
- Refactor inference code, separate evaluation and prediction.
- Related example PR https://github.com/second-state/WasmEdge-WASINN-examples/pull/97

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review bot](https://github.com/flows-network/github-pr-summary/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

File: plugins/wasi_nn/ggml.cpp [hunk 1]
```diff
@@ -522,7 +602,47 @@ Expect<ErrNo> setInput(WasiNNEnvironment &Env, uint32_t ContextId,
...
   const std::string Prompt(reinterpret_cast<char *>(Tensor.Tensor.data()),
                            Tensor.Tensor.size());
-  CxtRef.LlamaInputs = llama_tokenize(LlamaContext, Prompt, AddBos, true);
+  if (GraphRef.MMProjModelPath == ""sv || GraphRef.ImagePath == ""sv) {
+    // Text only prompt.
+    CxtRef.LlamaInputs = llama_tokenize(LlamaContext, Prompt, AddBos, true);
+  } else {
+    // Handle llava format prompt.
+    // We split prompt by {WASMEDGE_IMAGE} as placeholder and save the position.
+    const std::string_view PromptPlaceholder = "{WASMEDGE_IMAGE}";
```
Comment (hydai) [inline plugins/wasi_nn/ggml.cpp:611]: Let's use `<image>`, the same as llama.cpp

**Review by hydai — CHANGES_REQUESTED** [review]

File: plugins/wasi_nn/ggml.cpp (hunk 1 above)
Comment (dm4) [reply plugins/wasi_nn/ggml.cpp]: Updated.

File: plugins/wasi_nn/ggml.cpp [hunk 2]
```diff
@@ -813,133 +943,113 @@ Expect<ErrNo> computeSingle(WasiNNEnvironment &Env,
...
-          return ErrNo::RuntimeError;
-        }
-
-        CxtRef.LlamaNPast += NEval;
+      bool EvalImageStatus =
+          llava_eval_image_embed(CxtRef.LlamaContext, CxtRef.LlavaImageEmbd,
+                                 GraphRef.BatchSize, &CxtRef.LlamaNPast);
+      if (!EvalImageStatus) {
+        spdlog::error(
+            "[WASI-NN] GGML backend: failed to evaluate embed image tokens ."sv);
```
Comment (hydai) [inline plugins/wasi_nn/ggml.cpp:1002]: Extra space.

**Review by hydai — CHANGES_REQUESTED** [review]

File: plugins/wasi_nn/ggml.cpp (hunk 2 above)
Comment (dm4) [reply plugins/wasi_nn/ggml.cpp]: Fixed

**Review by hydai — APPROVED** [review]

**Pull request merged**